### PR TITLE
Test WeightNormalization layer can be saved in SavedModel format

### PR DESCRIPTION
**Pull request by SNeugberYoti**
Failing test to highlight issue #1788

### CODE REVIEW

**Comment by SNeugberYoti** [comment]
@googlebot I signed it!

**Comment by gabrieldemarmiesse** [comment]
Thank you @bhack  and @SNeugberYoti for the pull request with the test, that's very helpful. I'll use your branch and try to fix it.

**Comment by gabrieldemarmiesse** [comment]
@googlebot I consent.

**Comment by gabrieldemarmiesse** [comment]
So in the fix, we need to force the use of autograph with `@tf.function` so that we don't have to use any `tf.control_dependencies` after the assign.

**Comment by gabrieldemarmiesse** [comment]
I checked and my fix is wrong. `g` has no gradients. I don't think there is an easy way to fix this, to have something that works correctly we might deprecate this and give utility functions to plug into the layer instead. I don't believe that the WeightNormalization layer in this current form is a good idea. We'll have a regular flux of users complaining about bugs.

**Comment by SNeugberYoti** [comment]
Hi @gabrieldemarmiesse 

Looks like `g` has no gradients in your fix because the gradients of `assign` are undefined?

Maybe using `tf.custom_gradient` would work here? Similar to the solution here: https://github.com/tensorflow/tensorflow/issues/17735#issuecomment-437621701

**Comment by gabrieldemarmiesse** [comment]
@SNeugberYoti  thanks for the idea. The bigger issue here is the true value of the weights, and the gradients. I don't think the assignment is good at the semantic level, and it will confuse users who expect the kernel and `g` to be two different things. Now they're mixed together because of the assignment, which isn't correct, and will likely give use other problems moving forward. The right solution is a restructuring of the WeightNormalization layer. I lack time on my side. It would be very nice if a contributor could pick this up. I can give some guidance, but doing the pull request myself requires too much time.

**Comment by SNeugberYoti** [comment]
@gabrieldemarmiesse fair enough. I'd generally be happy to give this a go, if you could outline what you had in mind? I struggle to see a way around changing the kernel of the wrapped layer while also applying the weight normalisation at the right time.

**Comment by SNeugberYoti** [comment]
Regarding your comment [here](https://github.com/tensorflow/addons/pull/1797/files#r422347435), I'm not sure subclassing is the way forward, the wrapper class does seem like the correct approach to me.

> Now they're mixed together because of the assignment, which isn't correct, and will likely give use other problems moving forward. 

Technically this is correct though as this method fundamentally redefines the kernel. Going by the definition of the paper:

>  [...] We do so by expressing the weight vectors in terms of the new parameters using w = g v / ||v||


Going with the subclass approach, wouldn't this require the user to re-implement the "call" method of the base class and change the kernel there, which is basically what the wrapper is doing currently? The current approach seems more robust to me in that case.

I may equally just not know enough of the details though. Do you have an example of where you see the current approach running in to problems in the future?

Since @MokkeMeguru has been working on the layer for the other issue, maybe worth involving here as well.

**Comment by gabrieldemarmiesse** [comment]
For example, one huge problem we have with the implementation that I made in this pull requests is that the weights will change when doing a "model.predict()", because of the assign()

**Comment by SNeugberYoti** [comment]
Right, makes sense.

I have just subclassed Conv2D and added the weightnorm code in. It's quite messy and I can't see an easy way to pull the relevant parts out into utility functions, but it works. Gradients are fine and saving/loading the model is also fine. Especially the data-dependent initialisation made this tricky, having to run the layer once without the normalization functionality.

I wonder why just updating the kernel as in the original implementation (`self.layer.kernel = ...`) breaks model loading though, as I'm not doing much else in my subclassed implementation either.

**Comment by MokkeMeguru** [comment]
(I have one alternative idea. That is to create a wrapper class of **tf.keras.Layer** , and ask someone wants to use WeightNormalization to create a class such as WeightNorm-Conv2D inherit from its class. When I have a similar problem in [my Flow-based Model's Library](https://github.com/MokkeMeguru/TFGENZOO/blob/master/TFGENZOO/flows/flowbase.py#L10-L57), I used this solution. But this may be a misguided idea sry)

e.g.
```
class WeightNormLayer(tf.keras.layers.Layer):
    ...

class WeightNormedConv2D(WeightNormLayer):
    def build(input_shape):
        self.w = self.add_weightnormed_weight(name="weight_w", shape=self.filter_sizes)
```

**Comment by dathudeptrai** [comment]
what is a different betweent tf_addons's weightnorm and tfp's weightnorm here (https://github.com/tensorflow/probability/blob/master/tensorflow_probability/python/layers/weight_norm.py) ?

**Comment by bhack** [comment]
The reference paper Is the same https://arxiv.org/abs/1602.07868 Tim Salimans, Diederik P. Kingma (2016)

In TF probability it was introduced Aug 28, 2019 so It Is a more recent adoption then Addons but I've not check impl details.

**Comment by SNeugberYoti** [comment]
Implementation looks relatively similar, and suffers from what appears to be the same issue. Saving a model with their implementation works fine, but loading it again breaks.

**Comment by bhack** [comment]
I think It could confuse a little bit users cause probability It Is not a SIG and it is something ufficially supported by the Google Tensorflow team /cc @seanpmorgan

**Comment by seanpmorgan** [comment]
Thank you for your contribution. We sincerely apologize for any delay in reviewing, but TensorFlow Addons is transitioning to a minimal maintenance and release mode. New features will not be added to this repository. For more information, please see our public messaging on this decision:
[TensorFlow Addons Wind Down](https://github.com/tensorflow/addons/issues/2807)

Please consider sending feature requests / contributions to other repositories in the TF community with a similar charters to TFA:
[Keras](https://github.com/keras-team/keras)
[Keras-CV](https://github.com/keras-team/keras-cv)
[Keras-NLP](https://github.com/keras-team/keras-nlp)
